tests/unit/websocket_test.cc

```diff
@@ -31,6 +32,10 @@ SEASTAR_TEST_CASE(test_websocket_handshake) {
         auto output = sock.output();
 
         websocket::server dummy;
+        dummy.register_handler("echo", [] (input_stream<char>& in,
+                    output_stream<char>& out) {
+            return make_ready_future<>();
```

Comment:
If we pass connected_socket to the lambda, then we can share the handlers with other socket type handlers, and allow operations on sub-streams (like aborting a sub-stream).

Reply:
Passing an input & output stream pair was done here with a specific future use case in mind - adding a WebSocket server to Scylla and routing decoded WebSocket requests straight to the CQL server implementation (it would be cql-over-websocket). I think that streams offer a good abstraction, because the input data is not identical to raw data you read from the socket, it's a pipeline: read masked data from WebSocket -> unmask the data -> let the user handle unmasked data via reading from input stream -> let the user produce an unmasked response via writing to the output stream -> encode the response and send it.

Reply:
In other words, streams already hide unnecessary protocol details from the user - the handler operates on unmasked data and doesn't care what happens right after it's received from the socket or before it's sent.

Reply:
Of course it's layered, the question is what should be the abstraction.

connected_socket provides shutdown_input()/shutdown_output(), while pure streams don't have this ability (but can still shutdown the entire connection).

1. Does the websocket specification support async termination?
2. Is it useful? Or is it enough to terminate the entire outer connection?

shutdown is useful with an unresponsive peer, so maybe not.

Reply:
At least for the intended use case, terminating the whole connection on the first sight of any issue is perfectly acceptable - the user (e.g. an in-browser terminal) will just reconnect.